### DISCUSSION TITLE
If I reuse the same rpmtd for multiple headerGet() or rpmtdGetString() calls, would this cause a memory leak?

### DISCUSSION THREAD

**Discussion by shanebishop** [rpm-software-management/rpm#2492] [Q&A] [answered]
I would like help determining if I am incorrectly using `rpmtdFreeData()` and introducing a memory leak (more details are after my code). Here is my C code:
```c
// Compile with:
// gcc -Wall -Wextra -Werror -Og -g -lrpm -lrpmio mre.c -o e

#include <stdlib.h>
#include <stdio.h>
#include <rpm/rpmdb.h>
#include <rpm/rpmlib.h>
#include <rpm/rpmts.h>
#include <rpm/rpmmacro.h>

const char* getString(Header h, rpmTag tag, struct rpmtd_s* td) {
    if (headerGet(h, tag, td, HEADERGET_DEFAULT)) {
        return rpmtdGetString(td);
    }
    return NULL;
}

int main() {
    rpmts ts;
    Header h;
    rpmdbMatchIterator mi;
    rpmReadConfigFiles(NULL, NULL);

    ts = rpmtsCreate();
    if (!ts) {
        return 1;
    }

    mi = rpmtsInitIterator(ts, (rpmTag)RPMDBI_PACKAGES, NULL, 0);
    if (!mi) {
        rpmtsFree(ts);
        return 1;
    }

    h = rpmdbNextIterator(mi);
    if (h == NULL) {
        return 1;
    }

    struct rpmtd_s td;

    const char* s1 = getString(h, RPMTAG_NAME, &td);
    if (s1 != NULL)
        printf("Name: [%s]\n", s1);

    const char* s2 = getString(h, RPMTAG_VERSION, &td);
    if (s2 != NULL)
        printf("Version: [%s]\n", s2);

    // Note that I only free td once after using it twice
    rpmtdFreeData(&td);

    rpmdbFreeIterator(mi);
    rpmtsFree(ts);
    rpmFreeMacros(NULL);
    rpmFreeRpmrc();
    return 0;
}
```

Since I only free the `td` once, even though I use it twice, is it possible there is a memory leak?

I am worried that there might be a memory leak since there is an `xmalloc` [here](https://github.com/rpm-software-management/rpm/blob/rpm-4.18.x/lib/rpmtd.c#L12) and a free [here](https://github.com/rpm-software-management/rpm/blob/rpm-4.18.x/lib/rpmtd.c#L41) and [here](https://github.com/rpm-software-management/rpm/blob/rpm-4.18.x/lib/rpmtd.c#L44). I also wonder if in the stack for either `headerGet()` or `rpmtdGetString()` the `td` might have a new memory allocation.

If there is a memory leak in my code above, is the way to avoid a memory leak to use a separate `rpmtd`/`struct rpmtd_s` for each call to my `getString()` function?

**Comment by pmatilai** [comment, marked as answer]
It's okay to reuse the same struct rpmtd_s for multiple calls, BUT each headerGet() needs to be paired with a rpmtdFreeData(), otherwise it will leak memory. The worse problem in the above is that string pointers you get this way turn to garbage on rpmdbNextIterator() - it will not matter for the above snippet but will likely come bite you once it grows into a larger program.

HEADERGET_DEFAULT is a braindamaged mode which returns pointers into header memory for some data and allocated for others, unless you're very sure of what you're doing, HEADERGET_ALLOC is the right mode to use with headerGet().

If you only need non-array data from the header, you might as well save yourself some pain of awkward API's and just use headerGetAsString() to fetch the data converted into malloced string, plain and simple. For numeric data, headerGetNumber() is a similar shortcut.
